Profile line for 3d array added to profile.py

Maxyme opened the pull request:
I added a method to generate a profile line of a 3d array. It samples points around the direction axis, by rotating them along the same axis. The number of sample points is related to the line width. For a line width of 1 no rotation is needed. 
Otherwise, the number of angles is generated like this: rotation_angles = np.linspace(0, constants.pi, (linewidth \* 2) - 1) 

Also, arrays are converted to float before sending to ndi_map_coordinates, because the order value would be ignored when the arrays are int.
This is my first commit, let me know of any changes required. I also assume some functions may be optimized.

Maxyme commented:
@Korijn :
Fixed both the loop with transposed matrices and the main method with if/elif switch depending on dimension and multichannel.

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +122,222 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    perp_cols = np.array([np.linspace(col_i - col_width, col_i + col_width,
+                                      linewidth) for col_i in line_col])
+    perp_plane = np.array([np.linspace(slice_i - slice_width, slice_i + slice_width,
+                                       linewidth) for slice_i in line_slice])
```
We use (plane, row, column) as 3D coordinates, see [Coordinate conventions](http://scikit-image.org/docs/dev/user_guide/numpy_images.html#coordinate-conventions)

jni commented on skimage/measure/profile.py:
```diff
@@ -50,17 +53,27 @@ def profile_line(img, src, dst, linewidth=1,
-                  for i in range(img.shape[2])]
+    if img.ndim == 4 and multichannel:
+        # 3D RGB to be implemented
+        raise NotImplementedError('3D RGB to be implemented')
```
This should be trivial to implement following the example above? (Although you would have to stack the channels accordingly.)

jni replied:
(see `np.concatenate`)

jni replied:
@Maxyme do you think you could implement 3d multichannel following the example of 2D?

jni commented on skimage/measure/profile.py:
```diff
@@ -50,17 +53,27 @@ def profile_line(img, src, dst, linewidth=1,
-        pixels = ndi.map_coordinates(img, perp_lines,
-                                     order=order, mode=mode, cval=cval)
-    intensities = pixels.mean(axis=1)
+        raise ValueError('Invalid arguments')
```
Try to explain *why* they are invalid. e.g. `'profile_line is not implemented for images of dimension %s' % img.shape`

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +122,222 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    dst_row, dst_col, dst_dim3 = dst = np.asarray(dst, dtype=float)
+    d_row, d_col, d_plane = dst - src
+
+    # Get one unit vector perpendicular to direction vector to find a point
```
Can you try to explain this with a figure somewhere? I can't figure out what exactly you're doing here...

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +122,222 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+
+    # rotate all sample points around the direction axis if linewidth is > 1
+    if linewidth > 1:
+        perp_array = rotate_sample_points(linewidth, perp_array, src, dst)
```
If I understand correctly, you're taking a sampling vector (e.g. of length 3) and then rotating it around the centerline, to get a larger sampling vector representing a disk. But, wouldn't this method cause a denser sampling near the centerline compared to the edge of the cylinder?

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +122,222 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+        Width of the scan, perpendicular to the line,
+        since the line is 3d, this value is the diameter
+        of a 3d cylinder following the scan line.
+    perp_array, shape (3, N, C), float
```
What is C here?

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +122,222 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+
+    """
+
+    a = point_on_line[0]
```
`a, b, c = point_on_line`

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +122,222 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+
+    p1 = (a * (v ** 2 + w ** 2) - u * (b * v + c * w - u * x - v * y - w * z)) * (
+        1 - np.cos(angle_in_radians)) + x * np.cos(
+            angle_in_radians) + (-c * v + b * w - w * y + v * z) * np.sin(angle_in_radians)
```
Is there no way to simplify this? =\ At any rate, I think this is better with linebreaks aligned with the lower-priority operators:

```python
p1 = ((a * (v**2 + w**2) - u * (b*v + c*w + u*x - v*y - w*z)) *
      (1 - np.cos(angle_in_radians)) +
      (-c*v + b*w - w*y + v*z) * np.sin(angle_in_radians))
```
(and note that PEP8 prescribes breaking its operator spacing rules to denote execution priority.)

ivoflipse replied:
I'd imagine there's a vectorized version of this that doesn't require this unpacking

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +122,222 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+        1 - np.cos(angle_in_radians)) + z * np.cos(
+            angle_in_radians) + (-b * u + a * v - v * x + u * y) * np.sin(angle_in_radians)
+
+    return [p1, p2, p3]
```
Docstring specifies an array, so either this or the docstring needs to be changed.

jni commented:
@Maxyme Awesome, thank you for this contribution! It's something that's been on the back of my mind for a long time but my 3D geometry wasn't up to snuff. I've made some stylistic comments and a couple of implementation suggestions. After that I can have another stab at it. Thank you again!

jni commented:
To make some of the algebra less hairy, using quaternions might be handy. [This stackoverflow answer](http://stackoverflow.com/questions/4870393/rotating-coordinate-system-via-a-quaternion) has a nice, simple implementation that could be used here. Maybe? As I said, my 3D geometry is not great!

Maxyme commented:
@jni Thanks for your corrections. I'll have another go at it today. You are correct that the sampling method does generate a denser sampling near the center of the cylinder, or disk. One way of solving this is to reduce the angle size as we move farther from the center of the cylinder. Then, we could have a similar density of sample points.

Maxyme commented:
@jni I implemented some of the changes you suggested. Most importantly, I changed the method to increase the number of sample points as the distance between the point and the center increases.

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +122,211 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    The destination point is included in the profile, in contrast to
+    standard numpy indexing.
+    """
+    src_row, src_col, src_dim3 = src = np.asarray(src, dtype=float)
```
@Maxyme this still needs updating... Note that planes are the *leading* dimension, not the last one.

Maxyme replied:
@jni Ok, will do!

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +122,211 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+
+    line_col = np.linspace(src_col, dst_col, length)
+    line_row = np.linspace(src_row, dst_row, length)
+    line_plane = np.linspace(src_dim3, dst_dim3, length)
```
Optional: you can align the names by using `pln` to abbreviate `plane`.

Maxyme commented:
It seems to hang over numpy stack. 
http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.stack.html

Anyone has a clue why this happens?

ivoflipse commented:
np.stack got added in 1.10.0 (as that docs page says), while AppVeyor installs 1.8.1

Would it be an option to only provide this function if the user's numpy function is sufficient @jni?

Korijn commented:
skimage [requires `numpy>=1.6.1`](https://github.com/scikit-image/scikit-image/blob/master/requirements.txt), so you have to make sure your code can deal with that.

jni commented:
@Maxyme you can use np.concatenate together with np.newaxis to get the same effect.

Maxyme commented:
I changed the function to 
image3d = np.dstack((plane, plane, plane, plane, plane)).T
I think the dstack method was present in earlier versions.

ahojnnes commented:
@jni It seems you were pretty excited about this addition back in the time. The PR already looks very polished to me. It would be great, if we could merge it soon...

ahojnnes commented on skimage/measure/profile.py:
```diff
@@ -25,7 +25,9 @@ def profile_line(img, src, dst, linewidth=1,
-
+    multichannel : bool, optional
+        Whether the last axis of the image is to be interpreted as multiple
+        channels or another spatial dimension.
```
Add one empty line before "Returns"

ahojnnes commented on skimage/measure/profile.py:
```diff
@@ -50,17 +52,30 @@ def profile_line(img, src, dst, linewidth=1,
-        pixels = ndi.map_coordinates(img, perp_lines,
-                                     order=order, mode=mode, cval=cval)
-    intensities = pixels.mean(axis=1)
+        raise ValueError('profile_line is not implemented for images of dimension {0}'.format(img.shape))
```
Make sure to wrap all lines at column 80, here and elsewhere.

ahojnnes commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+
+
+def rotate_sample_points(linewidth, perp_array, src, dst):
+    """Return the evenly rotated coordinates of the sample points along a scan line in 3d
```
Add "." at end of line.

ahojnnes commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    return perp_array
+
+
+def rotate_sample_points(linewidth, perp_array, src, dst):
```
Since, this function will be public (no leading underscore), it would be good to have a more detailed description of what it does. Maybe you can add a short example.

jni replied:
@ahojnnes @Maxyme do we *want* this to be public?

ahojnnes commented:
@Maxyme Sorry for the long silence in this PR. I hope you are still interested in getting this PR merged into skimage!

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    # Get one unit vector perpendicular to direction vector to find a point
+    # that is one unit distance away from the destination vector
+    # (ex: ix + jy + kz = 0, then we can use x = y = 1)
+    # Try with z first if it is not 0, then the same for x, otherwise pick y
```
Three comments for this code block:

- Can you use "p, r, c" instead of "x, y, z" for the description? It's confusing to mix the two.
- You're repeating the same code while changing one thing... I think this could definitely be simplified to abstract away the "base" axis.
- At any rate, I'm confused about why we can assume that the width will be identical along two of the axes...? Can you explain this logic a bit further?

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    # (in contrast to standard numpy indexing)
+    length = np.ceil(np.linalg.norm([d_pln, d_row, d_col]) + 1)
+
+    line_col = np.linspace(src_col, dst_col, length)
```
Instead of adding 1 to length, use the `endpoint=True` keyword to `linspace`... Which is true by default anyway! But it's worth including explicitly in the function call anyway because of the ambiguity (which has resulted in an error in this case!).

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    line_pln = np.linspace(src_pln, dst_pln, length)
+
+    # find divisor to get only first half of array and center point if odd
+    first_half_index = np.int(np.ceil(np.float(linewidth)/2))
```
Remove call to `np.float` (which is the same as `float` anyway...). Instead, add `from __future__ import division` to the top of the file.

jni replied:
and similarly, `np.int` is the same as `int`, so simply use that.

jni replied:
Also, we can avoid the next conditional. So in the end I suggest:

```
first_half_index = max(1, int(np.ceil(linewidth / 2)))
```

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    perp_cols = np.array([np.linspace(col_i - col_width, col_i + col_width,
+                                      linewidth)[:first_half_index] for col_i in line_col])
+    perp_pln = np.array([np.linspace(slice_i - slice_width, slice_i + slice_width,
+                                       linewidth)[:first_half_index] for slice_i in line_pln])
```
Put `perp_pln` first! No point half-adhering to convention. =) Also, use `plns, rows, cols` or `pln, row, col`, for consistency. It'll make the code line up better. =)

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+        first_half_index = 1
+
+    perp_rows = np.array([np.linspace(row_i - row_width, row_i + row_width,
+                                      linewidth)[:first_half_index] for row_i in line_row])
```
Actually, instead of doing the indexing trick with `first_half_index` at all, you could do:

```
perp_rows = [np.linspace(row_i - row_width, row_i, np.ceil(linewidth / 2))
             for row_i in line_row]
```

Note also that you don't need the call to `np.array`. A single call, below with `perp_array`, will take care of all the nested lists.

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+        Width of the scan, perpendicular to the line,
+        since the line is 3d, this value is the diameter
+        of a 3d cylinder following the scan line.
+    perp_array, shape (3, M, N), float
```
Can you describe what `M` and `N` are here? I'm having trouble following this...

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    perp_array, shape (3, M, N), float
+        The coordinates of the profile along the scan line. The length of the
+        profile is the ceil of the computed length of the scan line.
+        The coordinates are 180 degrees apart.
```
180 or 90?

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    -------
+    sampling_array : array, shape (3, M, N), float
+        The coordinates of the 3d sample points along the scan line. The length of the
+        profile is the ceil of the computed length of the scan line.
```
If I understand correctly, N is the length of the profile, and M is the number of points sampled in each "disk" of the cylinder? Or vice versa... State this in the docstring, both for inputs and outputs.

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    return rotation_angles
+
+
+def _distance_point_line_3d(point, src, dst):
```
I suggest renaming this `_point_to_line_distance`. From my reading of the code, this code works in any dimension? It might be worth making it public.

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+        rotation_angles = np.array([0])
+    else:
+        rotation_angles = np.linspace(0, 2 * constants.pi, 2 * distance_point_line + 3)
+        rotation_angles = np.delete(rotation_angles, len(rotation_angles) - 1)
```
Is this not equal to `rotation_angles[:-1]`?

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+        The coordinates of the rotated point around the line
+    """
+    c, b, a = point_on_line
+    z, y, x = point_to_rotate
```
As before, don't use `z, y, x` where it can be avoided, instead use `p, r, c`.

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    z, y, x = point_to_rotate
+    w, v, u = unit_direction_vector
+
+    p1 = (a * (v ** 2 + w ** 2) - u * (b * v + c * w - u * x - v * y - w * z)) * \
```
PEP8 actually suggests breaking spacing to indicate operator priority. In the case of formulas like this one, this would be more readable, I think:

```
p1 = (a*(v**2 + w**2) - u*(b*v + c*w - u*x - v*y - w*z)) *
```

jni commented on skimage/measure/tests/test_profile.py:
```diff
@@ -3,7 +3,8 @@
 from skimage.measure import profile_line
 
-image = np.arange(100).reshape((10, 10)).astype(np.float)
+image = np.arange(100, dtype=np.float).reshape((10, 10))
```
👍 nicely done. =)

jni commented on skimage/measure/profile.py:
```diff
@@ -109,3 +124,229 @@ def _line_profile_coordinates(src, dst, linewidth=1):
+    sampling_array = []
+    for perp_points in perp_array.T:
+        rotation_angles = _rotation_angles_by_distance(dst, src, perp_points[0])
+        for angle in rotation_angles:  # the number of angles to use as rotation angles for the samping points
```
A slight modification:

```python
if angle == 0:
    points_array = perp_points
else:
    points_array = [_rotate_point_around_line(point, src, unit_direction_vector, angle)
                    for point in perp_points]
sampling_array.append(points_array)
```

jni commented:
@Maxyme As @ahojnnes, I apologise for the long delay in getting back to you. We are actively trying to improve our process to prevent such long waits in the future.

Do you have time soon to address our comments?

pep8speaks commented:
Hello @Maxyme! Thanks for updating the PR.

 - In the file [`skimage/measure/profile.py`](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py), following                are the PEP8 issues :
> [Line 67:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L67): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 72:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L72): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 77:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L77): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 82:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L82): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 84:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L84): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 79 characters)
> [Line 156:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L156): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)
> [Line 203:58](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L203): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 208:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L208): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 210:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L210): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 211:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L211): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 212:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L212): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 224:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L224): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 244:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L244): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 251:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L251): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)
> [Line 255:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L255): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 256:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L256): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 79 characters)
> [Line 258:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L258): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 262:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L262): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 79 characters)
> [Line 272:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L272): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 273:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L273): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 288:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L288): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 79 characters)
> [Line 318:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L318): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 79 characters)
> [Line 323:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L323): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 79 characters)
> [Line 347:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L347): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 351:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L351): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 355:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/profile.py#L355): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)


 - In the file [`skimage/measure/tests/test_profile.py`](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/tests/test_profile.py), following                are the PEP8 issues :
> [Line 115:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/tests/test_profile.py#L115): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 121:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/tests/test_profile.py#L121): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 127:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/tests/test_profile.py#L127): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)
> [Line 133:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/tests/test_profile.py#L133): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 139:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/tests/test_profile.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 145:80](https://github.com/Maxyme/scikit-image/blob/5f5723211f4287ea75483f159c4c7f397b3a16a7/skimage/measure/tests/test_profile.py#L145): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)

Maxyme commented on skimage/measure/_linalg.py:
```diff
@@ -0,0 +1,154 @@
+
+    p3 = (c * (u ** 2 + v ** 2) - w * (a * u + b * v - u * p - v * q - w * r)) * \
+         (1 - np.cos(angle_in_radians)) + r * np.cos(angle_in_radians) + \
+         (-b * u + a * v - v * p + u * q) * np.sin(angle_in_radians)
```
Hi, thanks for the comment, I'm in the long due process of updating this PR. I will add the reference here.

Maxyme commented:
@stefanv Yes, it's been too long. I've updated it a bit, but I think it still needs a thorough review. There may be duplicate or misplaced chunks of code (linalg, etc) so any suggestion in this regard is appreciated.
